tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -402,6 +413,329 @@ impl ServerKey {
+                    self.key.unchecked_scalar_add_assign(block, 1);
+                });
+
+            if compute_overflow {
+                self.key.unchecked_add_assign(
+                    &mut overflow_block,
+                    propagation_simulators.last().unwrap(),
+                );
+            }
+
+            (prepared_blocks, resolved_borrows)
+        };
+
+        let mut subtract_borrow_and_cleanup_prepared_blocks = || {
+            let message_extract_lut = self
+                .key
+                .generate_lookup_table(|block| (block >> 1) % self.message_modulus().0 as u64);
+
+            prepared_blocks
+                .par_iter_mut()
+                .enumerate()
+                .for_each(|(i, block)| {
+                    let grouping_index = i / grouping_size;
+                    let borrow = &resolved_borrows[grouping_index];
+                    crate::core_crypto::algorithms::lwe_ciphertext_sub_assign(
+                        &mut block.ct,
+                        &borrow.ct,
+                    );
+                    block.set_noise_level(block.noise_level() + borrow.noise_level());
+
+                    self.key
+                        .apply_lookup_table_assign(block, &message_extract_lut)
+                });
+        };
+
+        // Final step
+        if compute_overflow {
+            rayon::join(subtract_borrow_and_cleanup_prepared_blocks, || {
+                let borrow_flag_lut = self.key.generate_lookup_table(|block| (block >> 2) & 1);
+                self.key.unchecked_add_assign(
+                    &mut overflow_block,
+                    &resolved_borrows[resolved_borrows.len() - 1],
+                );
+                self.key
+                    .apply_lookup_table_assign(&mut overflow_block, &borrow_flag_lut);
+            });
+        } else {
+            subtract_borrow_and_cleanup_prepared_blocks();
+        }
+
+        lhs.clone_from_slice(&prepared_blocks);
+
+        if compute_overflow {
+            Some(BooleanBlock::new_unchecked(overflow_block))
+        } else {
+            None
+        }
+    }
+
+    /// blocks must e the result of (left_block - right_block + message_modulus)
+    /// (just like shortint::unchecked_sub_assign does on clean (no carries) ciphertext
+    fn compute_shifted_blocks_and_block_borrow_states(
+        &self,
+        blocks: &[Ciphertext],
+    ) -> (Vec<Ciphertext>, Vec<Ciphertext>) {
+        let num_blocks = blocks.len();
+
+        let message_modulus = self.message_modulus().0 as u64;
+
+        let block_modulus = self.message_modulus().0 * self.carry_modulus().0;
+        let num_bits_in_block = block_modulus.ilog2();
+
+        let grouping_size = num_bits_in_block as usize;
+
+        let shift_block_fn = |block| {
+            let overflow_guard = message_modulus;
+            let block = block % message_modulus;
+            (overflow_guard | block) << 1
+        };
```

Comment:
this works because the correcting term does not go beyon 2*msg_mod ?

Reply:
I remove the correcting term by doing ` block % message_modulus`, so add and overflow guard, to protect when ` block % message_modulus == 0`, as subtracting 1 would overflow the block and trash the padding bit

But actually, doing `block << 1` should also work as the block is the result from (`a -b + msg_mod`) so is already in a state where we can safely subtract a 1 with any provlem of overflow (same idea as in tfhe/src/integer/server_key/radix_parallel/sub.rs line 717, after `a - b + msg_mod` the result is in range `[1..=(2*msg_mod) -1]`


high bound: a=3,b=0,msg_mod=4 => (3-0 + 4) = 7 = (2*4) -1
low bound: a = 0, b=3,msg_mod=4 => (0 -3 + 4) = 1

So the lowest value after shortint_sub_assign is `1`, which means we can then safely subtract a borrow that may come from previous blocks